### DISCUSSION TITLE
Common test suite

### DISCUSSION THREAD

**Discussion by tstack** [gura-conf/gura#7] [Ideas (RFC)]
Are there plans to create a common suite of test configuration files?  It looks like the implementations have similar sets of files copied between them at the moment.

**Comment by Genarito** [comment]
Hi @tstack! 
First of all, thanks for your time opening this discussion.
Indeed, the tests were copied since the official implementations are a port of a parser that was made in Python in the first instance. How would you standardize a test suite for the different implementations?

**Reply by tstack** [reply]
> How would you standardize a test suite for the different implementations?

The TOML/compliance sounds like the right approach: have the implementations provide a driver program that takes an input file and then spits out the result; the test harness then takes care of running the driver with the input files and comparing the actual results to the expected.  You could even be prescriptive with error messages as well, so they were consistent across impls.  But, that might be pushing it.

**Reply by Genarito** [reply]
Great! I've just submitted a message to discuss this with the community. When I finish some priority things I'll get to work on this

**Reply by Genarito** [reply]
@tstack We have just published a new repo with all the available test cases for the current implementations. This repo will be introduced in the new version of Gura specs (2.0.0).

[Link to the Testing repo](https://github.com/gura-conf/testing)

**Comment by facundoq** [comment]
I think Tim refers to having a separate repository of gura (.gura) test
files that all implementations must be able to parse correctly.

Therefore if Gura changes (for example adding the empty keyword), and the
test files are changed accordingly, the tests for each implementation will
also start to fail.

This could be implemented as a git dependency.

**Reply by Genarito** [reply]
I see.. If I didn't misunderstand, It'd be something similar to [this TOML repo](https://github.com/toml-lang/compliance), wouldn't it?  We could make something similar

**Comment by Genarito** [comment]
Hi @tstack,

I have been thinking about this problem. The TOML proposal is good when the language specifications define only what is valid and what is not. In the case of Gura, there are standardized errors that are exceptions of the programming language in which the encoding/decoding is being implemented, so a language-agnostic solution is not possible. Also, TOML runs the encoder/decoder from the command line, so it is not useful for those implementations that are just libraries but do not have CLI functionality.

What I came up with, to save work to whoever is developing a Gura implementation, is to simply provide several `.ura` files separated in `valid` folders and the others in different folders representing the error they should throw. For example `ParseErrror`, `InvalidIndentationError`, etc.

This provides, optionally, several use cases that the developer can incorporate into the solution he/she is developing most conveniently.

What do you think about this?
